src/internal_telemetry/allocations/mod.rs

```diff
@@ -1,9 +1,53 @@
 //! Allocation tracking exposed via internal telemetry.
 
 mod allocator;
+use std::{
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Mutex,
+    },
+    thread,
+    time::Duration,
+};
+
+use arr_macro::arr;
+use metrics::gauge;
+use rand_distr::num_traits::ToPrimitive;
 
 use self::allocator::Tracer;
-pub(crate) use self::allocator::{AllocationGroupId, AllocationLayer, GroupedTraceableAllocator};
+
+pub(crate) use self::allocator::{
+    without_allocation_tracing, AllocationGroupId, AllocationLayer, GroupedTraceableAllocator,
+};
+
+use crossbeam_utils::CachePadded;
+
+const NUM_GROUPS: usize = 256;
+/// These arrays represent the allocations and deallocations for each group.
+/// We pad each Atomic to reduce false sharing effects.
+static GROUP_MEM_ALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
+    arr![CachePadded::new(AtomicU64::new(0)); 256];
+static GROUP_MEM_DEALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
+    arr![CachePadded::new(AtomicU64::new(0)); 256];
+
+// By using the Option type, we can do statics w/o the need of other creates such as lazy_static
+struct GroupInfo {
+    component_kind: Option<String>,
+    component_type: Option<String>,
+    component_id: Option<String>,
+}
+
+impl GroupInfo {
+    const fn new() -> Self {
+        Self {
+            component_id: None,
+            component_kind: None,
+            component_type: None,
+        }
+    }
+}
+
+static GROUP_INFO: [Mutex<GroupInfo>; NUM_GROUPS] = arr![Mutex::new(GroupInfo::new()); 256];
```

Comment:
Could this use one of the `ArcSwap` types to allow usage without locking? If we need locking, are they read-mostly, making a `RwLock` more suitable? Also, since these are all set and read at the same time, I recommend moving the `Option` out of the struct members and into the group info.

Reply:
Registration of an allocation group -- basically when a new component is created during config loading -- and getting the group information when reporting the metrics are the only two things that hit this data structure... which is why I explicitly asked Arshia to make this a `Mutex`, to focus on keeping it as simple as possible.

It's not read _or_ write-heavy, by any reasonable measure.